path_role.go

```diff
@@ -0,0 +1,433 @@
+	defer b.roleMutex.Unlock()
+
+	return nil, req.Storage.Delete(ctx, "role/"+strings.ToLower(roleName))
+}
+
+func (b *backend) pathRoleList(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	b.roleMutex.RLock()
+	defer b.roleMutex.RUnlock()
+
+	roles, err := req.Storage.List(ctx, "role/")
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(roles), nil
+}
+
+func (b *backend) pathRoleRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleEntry, err := b.lockedOCIRole(ctx, req.Storage, strings.ToLower(data.Get("role").(string)))
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: roleEntry.ToResponseData(),
+	}, nil
+}
+
+// create a Role
+func (b *backend) pathRoleCreate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	roleName := strings.ToLower(data.Get("role").(string))
+	if roleName == "" {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	if len(roleName) > MaxRoleNameLength {
+		return logical.ErrorResponse("role length exceeds the limit"), nil
+	}
+
+	validateRoleRegEx := regexp.MustCompile(`^[A-Za-z0-9]+$`).MatchString
+	if !validateRoleRegEx(roleName) {
+		return logical.ErrorResponse("role is invalid"), nil
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	roleEntry, err := b.nonLockedOCIRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	if roleEntry != nil {
+		return logical.ErrorResponse("The specified role already exists"), nil
+	}
+
+	roleEntry = &OCIRoleEntry{
+		Role:    roleName,
+		Version: 1,
+	}
+
+	if description, ok := data.GetOk("description"); ok {
+		roleEntry.Description = description.(string)
+	}
+
+	if add_ocid_list, ok := data.GetOk("add_ocid_list"); ok {
+		roleEntry.OcidList = add_ocid_list.([]string)
+		if len(roleEntry.OcidList) > MaxOCIDsPerRole {
+			return logical.ErrorResponse("Number of OCIDs for this role exceeds the limit"), nil
+		}
+	}
+
+	var resp logical.Response
+
+	ttl, ok := data.GetOk("ttl")
+	if ok {
+		ttlVal := ttl.(int)
+
+		if ttlVal > DefaultTTLSeconds {
+			return logical.ErrorResponse(fmt.Sprintf("Given ttl of %d seconds should be lesser than %d seconds;", ttlVal, DefaultTTLSeconds)), nil
+		}
+
+		if ttlVal < 0 {
+			return logical.ErrorResponse("ttl cannot be negative"), nil
+		}
+
+		roleEntry.TTL = ttlVal
+	} else {
+		roleEntry.TTL = DefaultTTLSeconds
+	}
+
+	if add_policy_list, ok := data.GetOk("add_policy_list"); ok {
+		roleEntry.PolicyList = add_policy_list.([]string)
+	}
+
+	if err := b.nonLockedSetOCIRole(ctx, req.Storage, roleName, roleEntry); err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
+}
+
+// Update if it already exists
+func (b *backend) pathRoleUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
```

Comment:
It usually possible, and very common, to have a single method that serves both Create and Update functions. Much of the logic is common, and at the start you might see something like:
```
	// Check if the role already exists
	role, err := b.role(ctx, req.Storage, roleName)
	if err != nil {
		return nil, err
	}

	// Create a new entry object if this is a CreateOperation
	if role == nil {
		if req.Operation == logical.UpdateOperation {
			return nil, errors.New("role entry not found during update operation")
		}
		role = new(RoleEntry)
	}
```

Reply:
Have updated the code